lib/pf/Authentication/Source/BillingSource.pm

```diff
@@ -0,0 +1,131 @@
+package pf::Authentication::Source::BillingSource;
+=head1 NAME
+
+pf::Authentication::Source::BillingSource
+
+=cut
+
+=head1 DESCRIPTION
+
+pf::Authentication::Source::BillingSource
+
+=cut
+
+use strict;
+use warnings;
+use Moose;
+use pf::config qw($FALSE $TRUE $default_pid $fqdn);
+use pf::Authentication::constants;
+use pf::util;
+
+extends 'pf::Authentication::Source';
+
+=head2 Attributes
+
+=head2 class
+
+=cut
+
+has '+class' => (default => 'abstact');
+
+has '+type' => (default => 'Billing');
+
+has '+unique' => (default => 1);
+
+has 'currency' => (is => 'rw', default => 'USD');
+
+has 'create_local_account' => (isa => 'Str', is => 'rw', default => 'no');
+
+has 'test_mode' => (is => 'rw', isa => 'Bool', default => 1);
+
+=head2 available_attributes
+
+Allow to make a condition on the user's email address.
+
+=cut
+
+sub available_attributes {
+  my $self = shift;
+
+  my $super_attributes = $self->SUPER::available_attributes;
+  my $own_attributes = [{ value => "username", type => $Conditions::SUBSTRING }];
+
+  return [@$super_attributes, @$own_attributes];
+}
+
+=head2 available_actions
+
+For an Null source, we limit the available actions to B<set role>, B<set access duration>, and B<set unreg date>.
+
+=cut
+
+sub available_actions {
+    return [ ];
+}
+
+=head2 match_in_subclass
+
+=cut
+
+sub match_in_subclass {
+    my ($self, $params, $rule, $own_conditions, $matching_conditions) = @_;
+    my $username =  $params->{'username'};
+    foreach my $condition (@{ $own_conditions }) {
+        if ($condition->{'attribute'} eq "username") {
+            if ( $condition->matches("username", $username) ) {
+                push(@{ $matching_conditions }, $condition);
+            }
+        }
+    }
+    return $username;
+}
+
+sub verify_url {
+    my ($self) = @_;
+    my $base_path = $self->_build_base_path;
+    return "$base_path/verify";
+}
+
+sub cancel_url {
+    my ($self) = @_;
+    my $base_path = $self->_build_base_path;
+    return "$base_path/cancel";
+}
+
+sub _build_base_path {
+    my ($self) = @_;
+    my $id     = $self->id;
+    my $base_path = "http://$fqdn/billing/$id";
```

Comment:
$fqdn is not always true.

When doing web auth it's the IP of the portal.

Should be handled differently